### PR TITLE
(Re-)Add Automated Market Maker docs

### PR DESCRIPTION
**Pull request by mDuo13**
Restores AMM reference docs that were removed by #1867 and #1874 now that AMM is part of the version 1.12.0 release. The English versions are updated to follow changes from [Ripple Open Source](https://opensource.ripple.com). The Japanese versions are not updated with the latest changes like AMMDelete, error code changes, and replacing the `lsfAMM` flag in AccountRoot with the `AMMID` field.

The "Create an AMM" tutorial should follow shortly in a separate PR.

### CODE REVIEW

File: content/concepts/tokens/automated-market-makers.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+labels:
+  - XRP
+  - Decentralized Exchange
+  - AMM
+---
+# Automated Market Makers
+
+_(Requires the [AMM amendment][] :not_enabled:)_
+
+Automated Market Makers (AMMs) are smart contracts that provide liquidity in the XRP Ledger's decentralized exchange. Each AMM holds a pool of two assets and enables users to swap between them at an exchange rate set by a formula.
```
Comment (gregtatcam) [inline content/concepts/tokens/automated-market-makers.md:15]: Why `smart contracts`?

File: content/concepts/tokens/automated-market-makers.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+For example, if you created an AMM with 5 ETH and 5 USD, and then someone exchanged 1.26 USD for 1 ETH, the pool now has 4 ETH and 6.26 USD in it. You can spend half your LP Tokens to withdraw 2 ETH and 3.13 USD.
+
+Anyone can deposit assets to an existing AMM. When they do, they receive new LP Tokens based on how much they deposited. The amount that a liquidity provider can withdraw from an AMM is based on the proportion of the AMM's LP Tokens they hold compared to the total number of LP Tokens outstanding.
+
+LP Tokens are like other tokens in the XRP Ledger, so you can use them in many [types of payments](payment-types.html), trade them in the decentralized exchange, or even deposit them as assets for new AMMs. (To receive LP Tokens as payment, you must set up a [trust line](trust-lines-and-issuing.html) with a nonzero limit with the AMM Account as the issuer.) However, you can _only_ send LP Tokens directly to the AMM (redeeming them) using the [AMMWithdraw][] transaction type, not through other types of payments. Similarly, you can only send assets to the AMM's pool through the [AMMDeposit][] transaction type.
```
Comment (gregtatcam) [inline content/concepts/tokens/automated-market-makers.md:46]: We don't allow AMMCreate with LPTokens.

File: content/concepts/tokens/automated-market-makers.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+### Auction Slot
+
+Unlike any previous Automated Market Makers, the XRP Ledger's AMM design has an _auction slot_ that a liquidity provider can bid on to get a discount on the trading fee for a 24-hour period. The bid must be paid in LP Tokens, which are returned to the AMM. No more than one account can hold the auction slot at a time, but the bidder can name up to 4 more accounts to also receive the discount. There is no minimum bid, but if the slot is currently occupied then you must outbid the current slot holder to displace them. If someone displaces you, you get part of your bid back depending on how much time remains. As long as you hold an active auction slot, you pay a discounted trading fee of 0% when making trades against that AMM.
```
Comment (gregtatcam) [inline content/concepts/tokens/automated-market-makers.md:65]: Discounted fee is tradingFee/10.

Comment (gregtatcam) [reply content/concepts/tokens/automated-market-makers.md:65]: LP Tokens left after reimbursing the previous owner are burnt.

File: content/concepts/tokens/automated-market-makers.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+- An [AMM entry][] describing the automated market maker itself.
+
+- A special [AccountRoot entry][] that issues the AMM's LP Tokens, and holds the AMM's XRP (if it has any).
+
+    The address of this AccountRoot is chosen somewhat randomly when the AMM is created, and it is different if the AMM is deleted and re-created. This is to prevent people from funding the AMM account with excess XRP in advance.
+
+- [Trust lines](trust-lines-and-issuing.html) to the special AMM Account for the tokens in the AMM's pool.
+
+These ledger entries are not owned by any account, so the [reserve requirement](reserves.html) does not apply to them. However, to prevent spam, the transaction to create an AMM has a special [transaction cost](transaction-cost.html) that requires the sender to burn a larger than usual amount of XRP.
+
```
Comment (gregtatcam) [inline content/concepts/tokens/automated-market-makers.md:83]: We added DirectoryNode entry to link AMM AccountRoot and AMM object to enable ledger_entry for amm type.

File: content/concepts/tokens/automated-market-makers.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+## Deletion
+
+An AMM is deleted when an [AMMWithdraw transaction][] withdraws all assets from its pool. This only happens by redeeming all of the AMM's outstanding LP Tokens. Deleting the AMM includes removes all the ledger entries associated with it, such as:
```
Comment (gregtatcam) [inline content/concepts/tokens/automated-market-makers.md:87]: Typo -`includes removes`?

File: content/concepts/tokens/automated-market-makers.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+- AMM
+- AccountRoot
+- Trust lines for the AMM's LP Tokens. Those trust lines would have a balance of 0 but may have other details, such as the limit, set to a non-default value.
+- Trust lines for the tokens that were in the AMM's pool.
+
+If there are more than 512 trust lines attached to the AMM account when it would be deleted, the withdraw succeeds and deletes as many trust lines as it can, but leaves the AMM in the ledger with no assets in its pool.
+
+While an AMM has no assets in its pool, anyone can delete it by sending an [AMMDelete transaction][]; if the remaining number of trust lines is still greater than the limit, multiple AMMDelete transactions may be necessary to fully delete the AMM. Alternatively, anyone can perform a [special deposit](ammdeposit.html#empty-amm-special-case) to fund the AMM as if it were new. No other operations are valid on an AMM with an empty asset pool.
+
```
Comment (gregtatcam) [inline content/concepts/tokens/automated-market-makers.md:97]: We also delete `DirectoryNode` AMM entry along with `AccountRoot` and `AMM` object.

File: content/references/http-websocket-apis/public-api-methods/ledger-methods/ledger_entry.md [hunk 7]
```diff
@@ -21,16 +21,17 @@ This method can retrieve several different types of data. You can select which t
...
 | `ledger_index`          | String or Unsigned Integer | _(Optional)_ The [ledger index][] of the ledger to use, or a shortcut string (e.g. "validated" or "closed" or "current") to choose a ledger automatically. (See [Specifying Ledgers][]) |
 
 The `generator` and `ledger` parameters are deprecated and may be removed without further notice.
 
-In addition to the general fields above, you must specify *exactly 1* of the following fields to indicate what type of object to retrieve, along with its sub-fields as appropriate. The valid fields are:
+In addition to the general fields above, you must specify *exactly 1* of the following fields to indicate what type of entry to retrieve, along with its sub-fields as appropriate. The valid fields are:
 
 - [`index`](#get-ledger-object-by-id)
 - [`account_root`](#get-accountroot-object)
+- [`amm`](#get-amm-object) :not_enabled:
```
Comment (gregtatcam) [inline content/references/http-websocket-apis/public-api-methods/ledger-methods/ledger_entry.md:34]: It's enabled now.

**Review by amarantha-k — APPROVED** [review]
Ran a spot check on the preview URL and it looks good

File: content/references/protocol-reference/ledger-data/ledger-object-types/amm.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    "TradingFee" : 600,
+    "VoteSlots" : [
+      {
+          "VoteEntry" : {
+            "Account" : "rJVUeRqDFNs2xqA7ncVE6ZoAhPUoaJJSQm",
+            "TradingFee" : 600,
+            "VoteWeight" : 100000
+          }
+      }
+    ]
```
Comment (gregtatcam) [inline content/references/protocol-reference/ledger-data/ledger-object-types/amm.md:66]: Should add `OwnerNode`.

File: content/references/protocol-reference/ledger-data/ledger-object-types/amm.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+| `AMMAccount`     | String              | AccountID         | Yes       | The address of the [special account](accountroot.html#special-amm-accountroot-objects) that holds this AMM's assets. |
+| `AuctionSlot`    | Object              | STObject          | No        | Details of the current owner of the auction slot, as an [Auction Slot object](#auction-slot-object). |
+| `LPTokenBalance` | [Currency Amount][] | Amount            | Yes       | The total outstanding balance of liquidity provider tokens from this AMM instance. The holders of these tokens can vote on the AMM's trading fee in proportion to their holdings, or redeem the tokens for a share of the AMM's assets which grows with the trading fees collected. |
+| `TradingFee`     | Number              | UInt16            | Yes       | The percentage fee to be charged for trades against this AMM instance, in units of 1/100,000. The maximum value is 1000, for a 1% fee. |
+| `VoteSlots`      | Array               | STArray           | No        | A list of vote objects, representing votes on the pool's trading fee. |
+
```
Comment (gregtatcam) [inline content/references/protocol-reference/ledger-data/ledger-object-types/amm.md:83]: Should add `OwnerNode`.

File: content/references/protocol-reference/transactions/transaction-results/tec-codes.md [hunk 10]
```diff
@@ -15,6 +15,13 @@ Transactions with `tec` codes destroy the XRP paid as a [transaction cost](trans
 
 | Code                       | Value | Explanation                             |
 |:---------------------------|:------|:----------------------------------------|
+| `tecAMM_ACCOUNT`           | 168   | The transaction failed because the operation is not allowed on Automated Market Maker (AMM) accounts. _(Requires the [AMM amendment][] :not_enabled:)_  |
+| `tecAMM_UNFUNDED`        | 162   | The [AMMCreate transaction][] failed because the sender does not have enough of the specified assets to fund it. _(Requires the [AMM amendment][] :not_enabled:)_  |
```
Comment (gregtatcam) [inline content/references/protocol-reference/transactions/transaction-results/tec-codes.md:19]: changed to `tecUNFUNDED_AMM`

Comment (gregtatcam) [inline content/references/protocol-reference/transactions/transaction-results/tec-codes.md:19]: Also `AMMDeposit`.

File: content/references/protocol-reference/transactions/transaction-results/tec-codes.md [hunk 11]
```diff
@@ -15,6 +15,13 @@ Transactions with `tec` codes destroy the XRP paid as a [transaction cost](trans
...
+| `tecAMM_BALANCE`         | 163   | The [AMMDeposit][] or [AMMWithdraw][] transaction failed because either the AMM or the user does not hold enough of one of the specified assets. (For example, you tried to withdraw more than the AMM holds.) _(Requires the [AMM amendment][] :not_enabled:)_  |
```
Comment (gregtatcam) [inline content/references/protocol-reference/transactions/transaction-results/tec-codes.md:20]: `AMMWithdraw` only.

File: content/references/protocol-reference/transactions/transaction-results/tec-codes.md [hunk 12]
```diff
@@ -15,6 +15,13 @@ Transactions with `tec` codes destroy the XRP paid as a [transaction cost](trans
...
+| `tecAMM_EMPTY`           | 166   | The AMM-related transaction failed because the AMM has no assets in its pool. In this state, you can only delete the AMM or fund it with a new deposit. _(Requires the [AMM amendment][] :not_enabled:)_  |
+| `tecAMM_FAILED`          | 164   | The AMM-related transaction failed. For [AMMDeposit][] or [AMMWithdraw][] this could be because the sender does not have enough of the specified assets, or the transaction requested an effective price that isn't possible with the available amounts. For [AMMBid][] this could be because the account does not have enough to win the bid or needs more than their specified maximum bid. For [AMMVote][], this could be because there are already too many votes from other accounts that hold more of this AMM's LP Tokens. _(Requires the [AMM amendment][] :not_enabled:)_  |
```
Comment (gregtatcam) [inline content/references/protocol-reference/transactions/transaction-results/tec-codes.md:22]: No longer used in `AMMVote`.

File: content/references/protocol-reference/transactions/transaction-types/ammbid.md [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+| `Asset2`       | Object              | STIssue           | Yes       | The definition for the other asset in the AMM's pool. In JSON, this is an object with `currency` and `issuer` fields (omit `issuer` for XRP). |
+| `BidMin`       | [Currency Amount][] | Amount            | No        | Pay at least this amount for the slot. Setting this value higher makes it harder for others to outbid you. If omitted, pay the minimum necessary to win the bid. |
+| `BidMax`       | [Currency Amount][] | Amount            | No        | Pay at most this amount for the slot. If the cost to win the bid is higher than this amount, the transaction fails. If omitted, pay as much as necessary to win the bid. |
+| `AuthAccounts` | Array               | STArray           | No        | A list of up to 4 additional accounts that you allow to trade at the discounted fee. This cannot include the address of the transaction sender. Each of these objects should be an [Auth Account object](#auth-account-objects). |
+
+You cannot specify both `BidMin` and `BidMax`.
```
Comment (gregtatcam) [inline content/references/protocol-reference/transactions/transaction-types/ammbid.md:65]: You can.

File: content/references/protocol-reference/transactions/transaction-types/ammbid.md [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+
+Like other "inner objects" that can appear in arrays, the JSON representation of each of these objects is wrapped in an object whose only key is the object type, `AuthAccount`.
+
+## Auction Slot Price
+
+If successful, the transaction automatically outbids the previous slot owner and debits the bid price from the sender's LP Tokens. The price to win the auction decreases over time, divided into 20 intervals of 72 minutes each. If the sender does not have enough LP Tokens to win the bid, or the price of the bid is higher than the transaction's `BidMax` value, the transaction fails with a `tecAMM_FAILED_BID` result.
+
+- If the auction slot is currently empty, expired, or in its last interval, the **minimum bid** is **0.001% of the AMM's total LP Tokens balance**.
+
+    **Caution:** This minimum value is a placeholder and may change before the AMM feature becomes finalized.
```
Comment (gregtatcam) [inline content/references/protocol-reference/transactions/transaction-types/ammbid.md:83]: Min slot price is `LPTokensBalance * tradingFee / 25`.

File: content/references/protocol-reference/transactions/transaction-types/ammbid.md [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+
+
+## Error Cases
+Besides errors that can occur for all transactions, {{currentpage.name}} transactions can result in the following [transaction result codes](transaction-results.html):
+
+| Error Code              | Description                                  |
+|:------------------------|:---------------------------------------------|
+| `tecAMM_EMPTY`          | The AMM has no assets in its pool. In this state, you can only delete the AMM or fund it with a new deposit. |
+| `tecAMM_FAILED`         | This transaction could not win the auction, either because the sender does not hold enough LP Tokens to pay the necessary bid or because the price to win the auction was higher than the transaction's specified `BidMax` value. |
+| `tecAMM_INVALID_TOKENS` | The sender of this transaction does not hold enough LP Tokens to meet the slot price. |
```
Comment (gregtatcam) [inline content/references/protocol-reference/transactions/transaction-types/ammbid.md:128]: Or the account is not `LP`.

File: content/references/protocol-reference/transactions/transaction-types/ammcreate.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+|:-------------|:--------------------|:------------------|:----------|:------------|
+| `Amount`     | [Currency Amount][] | Amount            | Yes       | The first of the two assets to fund this AMM with. This must be a positive amount. |
+| `Amount2`    | [Currency Amount][] | Amount            | Yes       | The second of the two assets to fund this AMM with. This must be a positive amount. |
+| `TradingFee` | Number              | UInt16            | Yes       | The fee to charge for trades against this AMM instance, in units of 1/100,000; a value of 1 is equivalent to 0.001%. The maximum value is `1000`, indicating a 1% fee. The minimum value is `0`. |
+
+One or both of `Amount` and `Amount2` can be [tokens](tokens.html); at most one of them can be [XRP](xrp.html). They cannot both have the same currency code and issuer. The tokens' issuers must have [Default Ripple](rippling.html#the-default-ripple-flag) enabled. If the [Clawback amendment][] :not_enabled: is enabled, those issuers must not have enabled the Allow Clawback flag. An AMM's LP tokens _can_ be used as one of the assets for another AMM.
```
Comment (gregtatcam) [inline content/references/protocol-reference/transactions/transaction-types/ammcreate.md:47]: Can't use AMM LP tokens as assets for another AMM.

File: content/references/protocol-reference/transactions/transaction-types/ammdelete.md [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+status: not_enabled
+---
+# AMMDelete
+[[Source]](https://github.com/XRPLF/rippled/blob/develop/src/ripple/app/tx/impl/AMMDelete.cpp "Source")
+
+_(Requires the [AMM amendment][] :not_enabled:)_
+
+Delete an empty [Automated Market Maker](automated-market-makers.html) (AMM) instance that could not be fully deleted automatically.
+
+Normally, an [AMMWithdraw transaction][] automatically deletes an AMM and all associated ledger entries when it withdraws all the assets from the AMM's pool. However, if there are too many trust lines to the AMM account to remove in one transaction, it may stop before fully removing the AMM. Similarly, an AMMDelete transaction removes up to a maximum of 512 trust lines; it may take several AMMDelete transactions to delete all the trust lines and the associated AMM. In all cases, only the last such transaction deletes the AMM and AccountRoot ledger entries.
```
Comment (gregtatcam) [inline content/references/protocol-reference/transactions/transaction-types/ammdelete.md:16]: And the `DirectoryNode` entry.

File: content/references/protocol-reference/transactions/transaction-types/ammdeposit.md [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+| `Asset`       | Object              | STIssue           | Yes       | The definition for one of the assets in the AMM's pool. In JSON, this is an object with `currency` and `issuer` fields (omit `issuer` for XRP). |
+| `Asset2`      | Object              | STIssue           | Yes       | The definition for the other asset in the AMM's pool. In JSON, this is an object with `currency` and `issuer` fields (omit `issuer` for XRP). |
+| `Amount`      | [Currency Amount][] | Amount            | No        | The amount of one asset to deposit to the AMM. If present, this must match the type of one of the assets (tokens or XRP) in the AMM's pool. |
+| `Amount2`     | [Currency Amount][] | Amount            | No        | The amount of another asset to add to the AMM. If present, this must match the type of the other asset in the AMM's pool and cannot be the same asset as `Amount`. |
+| `EPrice`      | [Currency Amount][] | Amount            | No        | The maximum effective price, in the deposit asset, to pay for each LP Token received. |
+| `LPTokenOut`  | [Currency Amount][] | Amount            | No        | How many of the AMM's LP Tokens to buy. |
+
```
Comment (gregtatcam) [inline content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:53]: `TradingFee` if `tfTwoAssetIfEmpty` is set.

File: content/references/protocol-reference/transactions/transaction-types/ammdeposit.md [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+
+### AMMDeposit Modes
+
+This transaction has five modes, defined by which flag you specify. Each mode expects a specific combination of fields. The modes fall into two categories:
```
Comment (gregtatcam) [inline content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:57]: Six modes.

File: content/references/protocol-reference/transactions/transaction-types/ammdeposit.md [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+| `tfLimitLPToken`    | `0x00400000` | `Amount` and `EPrice`      | Deposit up to the specified amount of one asset, but pay no more than the specified effective price per LP Token (after fees). |
+
+Any other combination of these fields and flags is invalid.
+
+
+### Single Asset Deposit Fee
+
+ The fee for a single asset deposit is calculated to be the same as if you had used the AMM to trade part of the deposit amount for the other asset, then done a double-asset deposit. The AMM's trading fee applies to the amount you would need to trade for, but not to the rest of the deposit. _For example, if the AMM's asset pool is split perfectly evenly between USD and EUR, and you try to deposit 100 USD, the amount of LP Tokens you receive is slightly less than if you had deposited 50 EUR + 50 USD, because you pay the trading fee to convert some of your USD to an equal amount of EUR._
+
+ The formula for how many LP Tokens you receive for a double-asset deposit is:
```
Comment (gregtatcam) [inline content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:85]: Shouldn't this be `single-asset`?

File: content/references/protocol-reference/transactions/transaction-types/ammdeposit.md [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+
+{{ include_svg("img/amm-single-asset-deposit-formula.svg", "L = T × ( (( 1 + (B - (F × (1 - W) × B)) ÷ P)^W) - 1)") }}
```
Comment (gregtatcam) [inline content/references/protocol-reference/transactions/transaction-types/ammdeposit.md:87]: The formula changed per Certik review request and is a lot more complicated.

**Review by gregtatcam — CHANGES_REQUESTED** [review]

**Pull request merged**